File: python/prism_sdk/autonomous_builtin_connectors.py
"""Credentialless built-in connector adapters for offline autonomous execution.

The connector registry and durable worker deliberately stop at a caller-owned executor
boundary.  That is the right security boundary for external providers, but it left the
Python SDK without a useful adapter for local development, deterministic evaluation, and
air-gapped deployments.  This module supplies that missing middle layer.

The built-in adapter accepts caller-supplied JSON metadata and returns a transient,
metadata-only observation.  It never opens a socket, discovers a provider, interprets a
credential, or claims that a local fixture is external evidence.  Its purpose is to make the
same exact operation contracts usable before an embedding application wires a real connector.

Every operation is digest-bound to one of the twelve autonomous domains.  The adapter projects
field names, shapes, counts, and SHA-256 digests; it never echoes the supplied values.  A
receipt journal therefore remains safe to persist, while the caller can still use the transient
observation to drive an offline planner, evaluator, replay harness, or local test.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from .authoring import content_digest
from .autonomous_connector_worker import (
    AutonomousConnectorOperationRegistry,
)
from .autonomous_connectors import (
    AutonomousConnectorObservation,
    AutonomousConnectorRegistration,
    AutonomousConnectorRegistry,
)
from .domain_evidence_provider_handoff import (
    DomainEvidenceProviderAuthPosture,
    DomainEvidenceProviderConnectorManifest,
)
from .domain_tools import (
    AUTONOMOUS_DOMAIN_NAMES,
    _identifier,
    _json_safe,
    _reject_secret_fields,
)
from .errors import ArgumentError


AUTONOMOUS_BUILTIN_CONNECTOR_SCHEMA = "bioprism-python-autonomous-builtin-connector/0.1"
AUTONOMOUS_BUILTIN_CONNECTOR_ID = "builtin.offline-evidence"
AUTONOMOUS_BUILTIN_CONNECTOR_VERSION = "1.0.0"
AUTONOMOUS_BUILTIN_CONNECTOR_PROVIDER = "local-offline"
MAX_AUTONOMOUS_BUILTIN_INPUT_BYTES = 2_000_000
MAX_AUTONOMOUS_BUILTIN_FIELDS = 128
MAX_AUTONOMOUS_BUILTIN_FIELD_NAME_BYTES = 256
MAX_AUTONOMOUS_BUILTIN_SEQUENCE_ITEMS = 1_024
MAX_AUTONOMOUS_BUILTIN_SHAPE_DEPTH = 16


# These are review prompts for an evaluator, not hard validation requirements.  The adapter
# can therefore ingest a sparse fixture and report ``partial`` instead of inventing evidence.
_RECOMMENDED_FIELDS: dict[str, tuple[str, ...]] = {
    "coding.repository_change_analysis": ("repository_digest", "changed_files", "test_results"),
    "browser.web_evidence_retrieval": ("source_digests", "retrieved_at", "citation_metadata"),
    "data.dataset_quality_profile": ("schema", "row_count", "column_count", "lineage"),
    "science.reproducible_evidence_acquisition": ("hypothesis", "evidence_digests", "analysis_digest"),
    "biomedical.clinical_data_review": ("provenance", "cohort_digest", "review_questions"),
    "neuroscience.signal_study_analysis": ("signal_digest", "sampling_rate", "study_design"),
    "operations.incident_runbook_observation": ("incident_digest", "telemetry_digest", "runbook_digest"),
    "enterprise.workflow_record_governance": ("workflow_digest", "record_type", "policy_digest"),
    "multi_agent.delegated_consensus_handoff": ("delegation_digest", "agent_digests", "conflicts"),
    "multimodal.asset_alignment": ("modalities", "asset_digests", "alignment_digest"),
    "cross_domain.evidence_fanout_synthesis": ("domain_digests", "evidence_digests", "route_digest"),
    "evaluation.benchmark_replay_analysis": ("benchmark_digest", "case_count", "replay_digest"),
}

_IDENTITY_FIELDS = frozenset({"operation_id", "subject_digest"})


def _digest(name: str, value: Any) -> str:
    if not isinstance(value, str) or len(value) != 64 or any(
        character not in "0123456789abcdef" for character in value
    ):
        raise ArgumentError(f"{name} must be a lowercase SHA-256 digest")
    return value


def _bounded_field_name(value: Any) -> str:
    if not isinstance(value, str) or not value.strip() or "\x00" in value:
        raise ArgumentError("built-in connector field names must be non-empty strings")
    if len(value.encode("utf-8")) > MAX_AUTONOMOUS_BUILTIN_FIELD_NAME_BYTES:
        raise ArgumentError("built-in connector field name exceeds its bound")
    return value


def _shape(value: Any, *, depth: int = 0) -> dict[str, Any]:
    """Return a bounded shape descriptor without retaining any input value."""

    if depth > MAX_AUTONOMOUS_BUILTIN_SHAPE_DEPTH:
        return {"type": "depth_limited"}
    if value is None:
        return {"type": "null"}
    if isinstance(value, bool):
        return {"type": "boolean"}
    if isinstance(value, str):
        return {"type": "string", "bytes": len(value.encode("utf-8"))}
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return {"type": "number"}
    if isinstance(value, Mapping):
        keys = tuple(str(key) for key in value)
        return {
            "type": "object",
            "field_count": len(keys),
            "field_names_digest": content_digest(sorted(keys)),
        }
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        return {
            "type": "array",
            "item_count": len(value),
            "item_types": sorted({
                _shape(item, depth=depth + 1).get("type", "unknown")
                for item in value[:MAX_AUTONOMOUS_BUILTIN_SEQUENCE_ITEMS]
            }),
        }
    return {"type": type(value).__name__}


def _content_present(value: Any) -> bool:
    if value is None or value == "":
        return False
    if isinstance(value, Mapping):
        return any(_content_present(child) for child in value.values())
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        return any(_content_present(child) for child in value)
    return True


def _field_projection(request: Mapping[str, Any]) -> tuple[dict[str, Any], ...]:
    fields: list[dict[str, Any]] = []
    for raw_name in sorted(request):
        name = _bounded_field_name(raw_name)
        if name in _IDENTITY_FIELDS:
            continue
        value = request[name]
        fields.append(
            {
                "name": name,
                "digest": content_digest(value),
                "shape": _shape(value),
                "present": _content_present(value),
            }
        )
    if len(fields) > MAX_AUTONOMOUS_BUILTIN_FIELDS:
        raise ArgumentError("built-in connector request contains too many fields")
    return tuple(fields)


class AutonomousBuiltinConnectorAdapter:
    """Deterministic local executor for the full built-in operation catalogue."""

    def __init__(
        self,
        operation_registry: AutonomousConnectorOperationRegistry | None = None,
        *,
        connector_id: str = AUTONOMOUS_BUILTIN_CONNECTOR_ID,
        version: str = AUTONOMOUS_BUILTIN_CONNECTOR_VERSION,
    ) -> None:
        self.operation_registry = operation_registry or AutonomousConnectorOperationRegistry()
        if not isinstance(self.operation_registry, AutonomousConnectorOperationRegistry):
            raise ArgumentError("built-in connector operation_registry is invalid")
        self.connector_id = _identifier("built-in connector connector_id", connector_id)
        self.version = _identifier("built-in connector version", version)
        missing = sorted(
            set(contract.operation_id for contract in self.operation_registry.operations())
            .difference(_RECOMMENDED_FIELDS)
        )
        if missing:
            raise ArgumentError("built-in connector has no field profile for: " + ", ".join(missing))

    @property
    def domains(self) -> tuple[str, ...]:
        return tuple(AUTONOMOUS_DOMAIN_NAMES)

    @property
    def capabilities(self) -> tuple[str, ...]:
        # The handoff manifest intentionally caps capability labels at 64. Preserve the first
        # (primary) capability of every operation, then fill the remaining budget
        # deterministically with secondary aliases. The operation registry remains the
        # authority for the complete vocabulary; this is only the wire-level projection.
        operations = self.operation_registry.operations()
        primary = tuple(dict.fromkeys(operation.capabilities[0] for operation in operations))
        secondary = tuple(
            capability
            for capability in sorted({item for operation in operations for item in operation.capabilities})
            if capability not in primary
        )
        return (primary + secondary)[:64]

    def manifest(self) -> DomainEvidenceProviderConnectorManifest:
        return DomainEvidenceProviderConnectorManifest(
            connector_id=self.connector_id,
            version=self.version,
            provider=AUTONOMOUS_BUILTIN_CONNECTOR_PROVIDER,
            connector_kind="provider_api",
            domains=self.domains,
            capabilities=self.capabilities,
            auth_posture=DomainEvidenceProviderAuthPosture(
                status="none",
                secret_refs=(),
                does_not_claim=(
                    "no external provider was contacted",
                    "caller-supplied metadata is not independently verified",
                    "no credential material is accepted or retained",
                ),
            ),
        )

    def execute(
        self,
        manifest: DomainEvidenceProviderConnectorManifest,
        request: Mapping[str, Any],
    ) -> AutonomousConnectorObservation:
        if not isinstance(manifest, DomainEvidenceProviderConnectorManifest):
            raise ArgumentError("built-in connector manifest is invalid")
        if manifest.connector_id != self.connector_id or manifest.version != self.version:
            raise ArgumentError("built-in connector manifest identity does not match the adapter")
        if not isinstance(request, Mapping):
            raise ArgumentError("built-in connector request must be an object")
        safe_request = _json_safe(
            "built-in connector request",
            dict(request),
            maximum=MAX_AUTONOMOUS_BUILTIN_INPUT_BYTES,
        )
        _reject_secret_fields(safe_request)
        operation_id = _identifier("built-in connector operation_id", safe_request.get("operation_id"))
        contract = self.operation_registry.resolve(operation_id)
        subject_digest = _digest("built-in connector subject_digest", safe_request.get("subject_digest"))
        fields = _field_projection(safe_request)
        # Presence is based on an explicit field, not truthiness. Empty conflict lists,
        # zero counts, and nullability decisions are meaningful caller evidence and must not
        # be silently relabeled as omitted input.
        available_fields = tuple(field["name"] for field in fields)
        recommended_fields = _RECOMMENDED_FIELDS[contract.operation_id]
        missing_fields = tuple(field for field in recommended_fields if field not in available_fields)
        has_evidence = any(field["present"] for field in fields)
        status = "observed" if has_evidence and not missing_fields else "partial"
        field_digests = {field["name"]: field["digest"] for field in fields}
        field_shapes = {field["name"]: field["shape"] for field in fields}
        value = {
            "schema": AUTONOMOUS_BUILTIN_CONNECTOR_SCHEMA,
            "operation_id": contract.operation_id,
            "domain": contract.domain,
            "subject_digest": subject_digest,
            "operation_digest": contract.operation_digest,
            "operation_capabilities": list(contract.capabilities),
            "evaluator_signals": list(contract.evaluator_signals),
            "recommended_fields": list(recommended_fields),
            "available_fields": list(available_fields),
            "missing_fields": list(missing_fields),
            "field_digests": field_digests,
            "field_shapes": field_shapes,
            "field_count": len(fields),
            "input_digest": content_digest(safe_request),
            "status": status,
            "failure_class": None if status == "observed" else "incomplete_local_fixture",
            "evidence_posture": "caller_supplied_metadata;offline_fixture;not_external_validation",
            "effect_posture": "read_only;no_network;no_provider_invocation",
            "retention": "transient_metadata_projection;receipt_retains_digest_only",
            "secret_material": "never_accepted_or_returned",
        }
        return AutonomousConnectorObservation(
            value=value,
            status=status,
            failure_class=value["failure_class"],
        )

    def __call__(
        self,
        manifest: DomainEvidenceProviderConnectorManifest,
        request: Mapping[str, Any],
    ) -> AutonomousConnectorObservation:
        return self.execute(manifest, request)


def builtin_autonomous_connector_registration(
    operation_registry: AutonomousConnectorOperationRegistry | None = None,
    *,
    connector_id: str = AUTONOMOUS_BUILTIN_CONNECTOR_ID,
    version: str = AUTONOMOUS_BUILTIN_CONNECTOR_VERSION,
    approval_required: bool = True,
) -> AutonomousConnectorRegistration:
    """Create the reviewed all-domain registration without mutating a registry."""

    if not isinstance(approval_required, bool):
        raise ArgumentError("built-in connector approval_required must be a boolean")
    adapter = AutonomousBuiltinConnectorAdapter(
        operation_registry,
        connector_id=connector_id,
        version=version,
    )
    return AutonomousConnectorRegistration(adapter.manifest(), adapter, approval_required=approval_required)


def register_builtin_autonomous_connectors(
    registry: AutonomousConnectorRegistry,
    operation_registry: AutonomousConnectorOperationRegistry | None = None,
    *,
    connector_id: str = AUTONOMOUS_BUILTIN_CONNECTOR_ID,
    version: str = AUTONOMOUS_BUILTIN_CONNECTOR_VERSION,
    approval_required: bool = True,
    replace: bool = False,
) -> AutonomousConnectorRegistration:
    """Register the deterministic all-domain adapter in a caller-owned connector registry."""

    if not isinstance(registry, AutonomousConnectorRegistry):
        raise ArgumentError("built-in connector registration requires an AutonomousConnectorRegistry")
    registration = builtin_autonomous_connector_registration(
        operation_registry,
        connector_id=connector_id,
        version=version,
        approval_required=approval_required,
    )
    registry.register(registration, replace=replace)
    return registration


__all__ = [
    "AUTONOMOUS_BUILTIN_CONNECTOR_SCHEMA",
    "AUTONOMOUS_BUILTIN_CONNECTOR_ID",
    "AUTONOMOUS_BUILTIN_CONNECTOR_VERSION",
    "AUTONOMOUS_BUILTIN_CONNECTOR_PROVIDER",
    "MAX_AUTONOMOUS_BUILTIN_INPUT_BYTES",
    "MAX_AUTONOMOUS_BUILTIN_FIELDS",
    "MAX_AUTONOMOUS_BUILTIN_FIELD_NAME_BYTES",
    "MAX_AUTONOMOUS_BUILTIN_SEQUENCE_ITEMS",
    "MAX_AUTONOMOUS_BUILTIN_SHAPE_DEPTH",
    "AutonomousBuiltinConnectorAdapter",
    "builtin_autonomous_connector_registration",
    "register_builtin_autonomous_connectors",
]

File: typescript/src/autonomous-builtin-connectors.ts
import { ArgumentError, isObject } from "./errors.js";
import { AUTONOMOUS_DOMAIN_NAMES, type AutonomousDomainName } from "./autonomous.js";
import {
  AutonomousConnectorObservation,
  AutonomousConnectorRegistration,
  AutonomousConnectorRegistry,
  AutonomousConnectorRuntime,
  type AutonomousConnectorExecutor,
  type AutonomousConnectorReceiptStore,
} from "./autonomous-connectors.js";
import { AutonomousConnectorOperationRegistry } from "./autonomous-connector-worker.js";
import { AutonomousConnectorOperationFacade } from "./autonomous-connector-facade.js";
import { canonicalJson, digestJsonSync } from "./tooling.js";
import type {
  DomainEvidenceProviderConnectorManifest,
  JsonObject,
  JsonValue,
} from "./types.js";

/**
 * Credentialless deterministic connector for local development, air-gapped deployments,
 * replay tests, and evaluator fixtures. It projects metadata shape and digests only; it
 * never opens a socket, consumes a credential, or represents caller metadata as external
 * evidence.
 */
export const AUTONOMOUS_BUILTIN_CONNECTOR_SCHEMA = "bioprism-typescript-autonomous-builtin-connector/0.1" as const;
export const AUTONOMOUS_BUILTIN_CONNECTOR_ID = "builtin.offline-evidence" as const;
export const AUTONOMOUS_BUILTIN_CONNECTOR_VERSION = "1.0.0" as const;
export const AUTONOMOUS_BUILTIN_CONNECTOR_PROVIDER = "local-offline" as const;
export const MAX_AUTONOMOUS_BUILTIN_INPUT_BYTES = 2_000_000;
export const MAX_AUTONOMOUS_BUILTIN_FIELDS = 128;
export const MAX_AUTONOMOUS_BUILTIN_FIELD_NAME_BYTES = 256;
export const MAX_AUTONOMOUS_BUILTIN_SEQUENCE_ITEMS = 1_024;
export const MAX_AUTONOMOUS_BUILTIN_SHAPE_DEPTH = 16;

const RECOMMENDED_FIELDS: Readonly<Record<string, readonly string[]>> = {
  "coding.repository_change_analysis": ["repository_digest", "changed_files", "test_results"],
  "browser.web_evidence_retrieval": ["source_digests", "retrieved_at", "citation_metadata"],
  "data.dataset_quality_profile": ["schema", "row_count", "column_count", "lineage"],
  "science.reproducible_evidence_acquisition": ["hypothesis", "evidence_digests", "analysis_digest"],
  "biomedical.clinical_data_review": ["provenance", "cohort_digest", "review_questions"],
  "neuroscience.signal_study_analysis": ["signal_digest", "sampling_rate", "study_design"],
  "operations.incident_runbook_observation": ["incident_digest", "telemetry_digest", "runbook_digest"],
  "enterprise.workflow_record_governance": ["workflow_digest", "record_type", "policy_digest"],
  "multi_agent.delegated_consensus_handoff": ["delegation_digest", "agent_digests", "conflicts"],
  "multimodal.asset_alignment": ["modalities", "asset_digests", "alignment_digest"],
  "cross_domain.evidence_fanout_synthesis": ["domain_digests", "evidence_digests", "route_digest"],
  "evaluation.benchmark_replay_analysis": ["benchmark_digest", "case_count", "replay_digest"],
};

const IDENTITY_FIELDS = new Set(["operation_id", "subject_digest"]);
const SECRET_FIELD_MARKERS = new Set([
  "apikey", "authorization", "bearer", "credential", "credentials", "password", "secret", "secretkey",
  "token", "accesstoken", "refreshtoken", "privatekey", "clientsecret", "gsk", "sk",
]);

function bytes(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}

function rejectsSecretField(name: string): boolean {
  const normalized = name.toLowerCase().replace(/[^a-z0-9]/g, "");
  return SECRET_FIELD_MARKERS.has(normalized) || normalized.startsWith("gsk") || normalized.startsWith("skproj");
}

function safeInput(name: string, value: unknown, maximum: number, depth = 0): JsonValue {
  if (depth > 32) throw new ArgumentError(`${name} is too deeply nested`);
  if (value === null || typeof value === "string" || typeof value === "boolean") return value;
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (Array.isArray(value)) {
    const result = value.map((item) => safeInput(name, item, maximum, depth + 1));
    if (bytes(canonicalJson(result)) > maximum) throw new ArgumentError(`${name} exceeds ${maximum} bytes`);
    return result;
  }
  if (isObject(value)) {
    const result: JsonObject = {};
    for (const [key, child] of Object.entries(value)) {
      if (rejectsSecretField(key)) throw new ArgumentError(`${name} contains credential-shaped fields`);
      result[key] = safeInput(name, child, maximum, depth + 1);
    }
    if (bytes(canonicalJson(result)) > maximum) throw new ArgumentError(`${name} exceeds ${maximum} bytes`);
    return result;
  }
  throw new ArgumentError(`${name} must be JSON-safe`);
}

function identifier(name: string, value: unknown): string {
  if (typeof value !== "string" || !value || value.length > 256 || !/^[A-Za-z0-9_.:-]+$/.test(value)) throw new ArgumentError(`${name} is outside its identifier contract`);
  return value;
}

function digest(name: string, value: unknown): string {
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) throw new ArgumentError(`${name} must be a lowercase SHA-256 digest`);
  return value;
}

function boundedFieldName(value: unknown): string {
  if (typeof value !== "string" || !value.trim() || value.includes("\u0000") || bytes(value) > MAX_AUTONOMOUS_BUILTIN_FIELD_NAME_BYTES) throw new ArgumentError("built-in connector field names are outside their bound");
  return value;
}

function shape(value: JsonValue, depth = 0): JsonObject {
  if (depth > MAX_AUTONOMOUS_BUILTIN_SHAPE_DEPTH) return { type: "depth_limited" };
  if (value === null) return { type: "null" };
  if (typeof value === "boolean") return { type: "boolean" };
  if (typeof value === "string") return { type: "string", bytes: bytes(value) };
  if (typeof value === "number") return { type: "number" };
  if (Array.isArray(value)) {
    return {
      type: "array",
      item_count: value.length,
      item_types: [...new Set(value.slice(0, MAX_AUTONOMOUS_BUILTIN_SEQUENCE_ITEMS).map((item) => String(shape(item, depth + 1).type)))].sort(),
    };
  }
  return { type: "object", field_count: Object.keys(value).length, field_names_digest: digestJsonSync(Object.keys(value).sort()) };
}

function contentPresent(value: JsonValue): boolean {
  if (value === null || value === "") return false;
  if (Array.isArray(value)) return value.some((item) => contentPresent(item));
  if (isObject(value)) return Object.values(value).some((item) => contentPresent(item as JsonValue));
  return true;
}

function fieldProjection(request: JsonObject): JsonObject[] {
  const fields: JsonObject[] = [];
  for (const name of Object.keys(request).sort()) {
    const bounded = boundedFieldName(name);
    if (IDENTITY_FIELDS.has(bounded)) continue;
    const value = request[bounded];
    if (value === undefined) throw new ArgumentError("built-in connector request contains an undefined field");
    fields.push({ name: bounded, digest: digestJsonSync(value), shape: shape(value), present: contentPresent(value) });
  }
  if (fields.length > MAX_AUTONOMOUS_BUILTIN_FIELDS) throw new ArgumentError("built-in connector request contains too many fields");
  return fields;
}

function domainValue(value: string | undefined): AutonomousDomainName | null {
  if (value === undefined) return null;
  if (!AUTONOMOUS_DOMAIN_NAMES.includes(value as AutonomousDomainName)) throw new ArgumentError("built-in connector domain is unsupported");
  return value as AutonomousDomainName;
}

export class AutonomousBuiltinConnectorAdapter {
  readonly operationRegistry: AutonomousConnectorOperationRegistry;
  readonly connector_id: string;
  readonly version: string;
  readonly domain: AutonomousDomainName | null;

  constructor(options: { operationRegistry?: AutonomousConnectorOperationRegistry; connectorId?: string; version?: string; domain?: AutonomousDomainName } = {}) {
    this.operationRegistry = options.operationRegistry ?? new AutonomousConnectorOperationRegistry();
    if (!(this.operationRegistry instanceof AutonomousConnectorOperationRegistry)) throw new ArgumentError("built-in connector operationRegistry is invalid");
    this.connector_id = identifier("built-in connector connectorId", options.connectorId ?? AUTONOMOUS_BUILTIN_CONNECTOR_ID);
    this.version = identifier("built-in connector version", options.version ?? AUTONOMOUS_BUILTIN_CONNECTOR_VERSION);
    this.domain = domainValue(options.domain);
    const contracts = this.domain === null ? this.operationRegistry.operations() : this.operationRegistry.forDomain(this.domain);
    if (!contracts.length) throw new ArgumentError("built-in connector domain has no operation contract");
    const missing = contracts.map((contract) => contract.operation_id).filter((id) => RECOMMENDED_FIELDS[id] === undefined);
    if (missing.length) throw new ArgumentError(`built-in connector has no field profile for: ${missing.join(", ")}`);
  }

  get domains(): AutonomousDomainName[] {
    return this.domain === null ? [...AUTONOMOUS_DOMAIN_NAMES] : [this.domain];
  }

  get capabilities(): string[] {
    const contracts = this.domain === null ? this.operationRegistry.operations() : this.operationRegistry.forDomain(this.domain);
    const primary = [...new Set(contracts.map((operation) => operation.capabilities[0]).filter((value): value is string => value !== undefined))];
    const secondary = [...new Set(contracts.flatMap((operation) => operation.capabilities))].filter((value) => !primary.includes(value)).sort();
    return [...primary, ...secondary].slice(0, 128);
  }

  manifest(): DomainEvidenceProviderConnectorManifest {
    return {
      schema: "bioprism-devplat-domain-evidence-provider-connector-manifest/0.1",
      connector_id: this.connector_id,
      version: this.version,
      provider: AUTONOMOUS_BUILTIN_CONNECTOR_PROVIDER,
      connector_kind: "provider_api",
      domains: this.domains,
      capabilities: this.capabilities,
      transport: "caller_managed",
      auth_posture: {
        status: "none",
        secret_refs: [],
        does_not_claim: [
          "no external provider was contacted",
          "caller-supplied metadata is not independently verified",
          "no credential material is accepted or retained",
        ],
      },
    };
  }

  execute(manifest: DomainEvidenceProviderConnectorManifest, request: JsonObject): AutonomousConnectorObservation {
    if (!isObject(manifest) || manifest.connector_id !== this.connector_id || manifest.version !== this.version) throw new ArgumentError("built-in connector manifest identity does not match the adapter");
    const safeRequest = safeInput("built-in connector request", request, MAX_AUTONOMOUS_BUILTIN_INPUT_BYTES);
    if (!isObject(safeRequest)) throw new ArgumentError("built-in connector request must be an object");
    const operationId = identifier("built-in connector operation_id", safeRequest.operation_id);
    const contract = this.operationRegistry.resolve(operationId);
    if (this.domain !== null && contract.domain !== this.domain) throw new ArgumentError("built-in connector operation exceeds its domain scope");
    const subjectDigest = digest("built-in connector subject_digest", safeRequest.subject_digest);
    const fields = fieldProjection(safeRequest);
    const available = fields.map((field) => String(field.name));
    const recommended = RECOMMENDED_FIELDS[contract.operation_id] ?? [];
    const missing = recommended.filter((field) => !available.includes(field));
    const hasEvidence = fields.some((field) => field.present === true);
    const status = hasEvidence && missing.length === 0 ? "observed" : "partial";
    const fieldDigests = Object.fromEntries(fields.map((field) => [String(field.name), String(field.digest)]));
    const fieldShapes = Object.fromEntries(fields.map((field) => [String(field.name), field.shape as JsonValue]));
    const value: JsonObject = {
      schema: AUTONOMOUS_BUILTIN_CONNECTOR_SCHEMA,
      operation_id: contract.operation_id,
      domain: contract.domain,
      subject_digest: subjectDigest,
      operation_digest: contract.operation_digest,
      operation_capabilities: [...contract.capabilities],
      evaluator_signals: [...contract.evaluator_signals],
      recommended_fields: [...recommended],
      available_fields: available,
      missing_fields: missing,
      field_digests: fieldDigests,
      field_shapes: fieldShapes,
      field_count: fields.length,
      input_digest: digestJsonSync(safeRequest),
      status,
      failure_class: status === "observed" ? null : "incomplete_local_fixture",
      evidence_posture: "caller_supplied_metadata;offline_fixture;not_external_validation",
      effect_posture: "read_only;no_network;no_provider_invocation",
      retention: "transient_metadata_projection;receipt_retains_digest_only",
      secret_material: "never_accepted_or_returned",
    };
    return new AutonomousConnectorObservation(value, status, status === "observed" ? null : "incomplete_local_fixture");
  }

  asExecutor(): AutonomousConnectorExecutor {
    return (manifest, request) => this.execute(manifest, request);
  }
}

export function builtinAutonomousConnectorRegistration(options: {
  operationRegistry?: AutonomousConnectorOperationRegistry;
  connectorId?: string;
  version?: string;
  approvalRequired?: boolean;
  domain?: AutonomousDomainName;
} = {}): AutonomousConnectorRegistration {
  if (options.approvalRequired !== undefined && typeof options.approvalRequired !== "boolean") throw new ArgumentError("built-in connector approvalRequired must be boolean");
  const adapter = new AutonomousBuiltinConnectorAdapter(options);
  return new AutonomousConnectorRegistration(adapter.manifest(), adapter.asExecutor(), options.approvalRequired ?? true);
}

export function registerBuiltinAutonomousConnectors(
  registry: AutonomousConnectorRegistry,
  options: { operationRegistry?: AutonomousConnectorOperationRegistry; connectorId?: string; version?: string; approvalRequired?: boolean; replace?: boolean } = {},
): AutonomousConnectorRegistration {
  if (!(registry instanceof AutonomousConnectorRegistry)) throw new ArgumentError("built-in connector registration requires an AutonomousConnectorRegistry");
  const registration = builtinAutonomousConnectorRegistration(options);
  registry.register(registration, { replace: options.replace === true });
  return registration;
}

export function builtinAutonomousDomainConnectorRegistrations(options: {
  operationRegistry?: AutonomousConnectorOperationRegistry;
  connectorId?: string;
  version?: string;
  approvalRequired?: boolean;
} = {}): AutonomousConnectorRegistration[] {
  const operationRegistry = options.operationRegistry ?? new AutonomousConnectorOperationRegistry();
  return AUTONOMOUS_DOMAIN_NAMES.map((domain) => builtinAutonomousConnectorRegistration({
    operationRegistry,
    connectorId: `${options.connectorId ?? AUTONOMOUS_BUILTIN_CONNECTOR_ID}.${domain}`,
    version: options.version,
    approvalRequired: options.approvalRequired,
    domain,
  }));
}

export function registerBuiltinAutonomousDomainConnectors(
  registry: AutonomousConnectorRegistry,
  options: { operationRegistry?: AutonomousConnectorOperationRegistry; connectorId?: string; version?: string; approvalRequired?: boolean; replace?: boolean } = {},
): AutonomousConnectorRegistration[] {
  if (!(registry instanceof AutonomousConnectorRegistry)) throw new ArgumentError("built-in domain connector registration requires an AutonomousConnectorRegistry");
  const registrations = builtinAutonomousDomainConnectorRegistrations(options);
  const existing = new Set(registry.registrations().map((registration) => registration.connector_id));
  if (options.replace !== true && registrations.some((registration) => existing.has(registration.connector_id))) throw new ArgumentError("a built-in domain connector is already registered");
  if (existing.size + registrations.filter((registration) => !existing.has(registration.connector_id)).length > 256) throw new ArgumentError("built-in domain connector registration exceeds registry capacity");
  for (const registration of registrations) registry.register(registration, { replace: options.replace === true });
  return registrations;
}

/**
 * Build an immediately usable local connector runtime for an embedding application.
 * The returned runtime still requires the normal reviewed selection and approval gates;
 * this helper only removes boilerplate from keyless development and evaluation setups.
 */
export function createBuiltinAutonomousConnectorRuntime(options: {
  domainScoped?: boolean;
  operationRegistry?: AutonomousConnectorOperationRegistry;
  connectorId?: string;
  version?: string;
  approvalRequired?: boolean;
  receiptStore?: AutonomousConnectorReceiptStore;
  receiptSink?: (receipt: unknown) => unknown | Promise<unknown>;
} = {}): {
  operationRegistry: AutonomousConnectorOperationRegistry;
  registry: AutonomousConnectorRegistry;
  runtime: AutonomousConnectorRuntime;
  operationFacade: AutonomousConnectorOperationFacade;
  registrations: AutonomousConnectorRegistration[];
} {
  const operationRegistry = options.operationRegistry ?? new AutonomousConnectorOperationRegistry();
  const registry = new AutonomousConnectorRegistry();
  const registrations = options.domainScoped === false
    ? [builtinAutonomousConnectorRegistration({ operationRegistry, connectorId: options.connectorId, version: options.version, approvalRequired: options.approvalRequired })]
    : builtinAutonomousDomainConnectorRegistrations({ operationRegistry, connectorId: options.connectorId, version: options.version, approvalRequired: options.approvalRequired });
  for (const registration of registrations) registry.register(registration);
  const runtime = new AutonomousConnectorRuntime(registry, {
    receiptStore: options.receiptStore,
    receiptSink: options.receiptSink,
  });
  const operationFacade = new AutonomousConnectorOperationFacade({ registry, runtime, operationRegistry });
  return { operationRegistry, registry, runtime, operationFacade, registrations };
}
